vmm/src/device_manager.rs

```diff
@@ -606,28 +607,38 @@ impl DeviceRelocation for AddressManager {
             PciBarRegionType::Memory32BitRegion | PciBarRegionType::Memory64BitRegion => {
                 // Update system allocator
                 if region_type == PciBarRegionType::Memory32BitRegion {
-                    self.allocator
-                        .lock()
-                        .unwrap()
-                        .free_mmio_hole_addresses(GuestAddress(old_base), len as GuestUsize);
+                    // Find the specific allocator that this BAR was allocated from and use it for new one
+                    for allocator in &self.pci_mmio32_allocators {
+                        let allocator_base = allocator.lock().unwrap().base();
+                        let allocator_end = allocator.lock().unwrap().end();
 
-                    self.allocator
-                        .lock()
-                        .unwrap()
-                        .allocate_mmio_hole_addresses(
-                            Some(GuestAddress(new_base)),
-                            len as GuestUsize,
-                            Some(len),
-                        )
-                        .ok_or_else(|| {
-                            io::Error::new(
-                                io::ErrorKind::Other,
-                                "failed allocating new 32 bits MMIO range",
-                            )
-                        })?;
+                        if old_base >= allocator_base.0 && old_base <= allocator_end.0 {
+                            allocator
+                                .lock()
+                                .unwrap()
+                                .free(GuestAddress(old_base), len as GuestUsize);
+
+                            allocator
+                                .lock()
+                                .unwrap()
+                                .allocate(
+                                    Some(GuestAddress(new_base)),
+                                    len as GuestUsize,
+                                    Some(len),
+                                )
+                                .ok_or_else(|| {
+                                    io::Error::new(
+                                        io::ErrorKind::Other,
+                                        "failed allocating new 32 bits MMIO range",
+                                    )
+                                })?;
+
+                            break;
+                        }
+                    }
```

Comment:
We are now doing the same thing for 32-bit and 64-bit MMIO regions, so we can further simplify the code like:

```
                let allocators = if region_type == PciBarRegionType::Memory32BitRegion {                                                                                                                      
                    &self.pci_mmio32_allocators                                                                                                                                                               
                } else {                                                                                                                                                                                      
                    &self.pci_mmio64_allocators                                                                                                                                                               
                };   

                // Find the specific allocator that this BAR was allocated from and use it for new one                                                                                                        
                for allocator in allocators { 
```

Reply:
This is a lot less repetitive. Updated.